src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeExceptionHelpers.cs

```diff
@@ -235,71 +224,47 @@ internal static void FailFast(string message, Exception? exception, RhFailFastRe
-                        var cookieSpan = new Span<byte>(pExceptionRecord->Cookie, CrashDumpRecord.CookieSize);
-                        // Random 10 bytes to identify the record
-                        ((ReadOnlySpan<byte>)new byte[] { 0x1c, 0x73, 0xd0, 0x2d, 0xda, 0x6b, 0x4c, 0xef, 0xbf, 0xa1 }).CopyTo(cookieSpan);
-                        "NETRUNTIME"u8.CopyTo(cookieSpan.Slice(10));
-                        pExceptionRecord->Type = 1;
-                        pExceptionRecord->Data = (void*)exceptionStack.GetPinnableReference();
-                        pExceptionRecord->Length = length;
-                    }
+                    _crashInfo.WriteExceptionInfo(exception);
                 }
             }
 
-#if TARGET_WINDOWS
-            uint errorCode = 0x80004005; // E_FAIL
-            // To help enable testing to bucket the failures we choose one of the following as errorCode:
-            // * hashcode of EETypePtr if it is an unhandled managed exception
-            // * HRESULT, if available
-            // * RhFailFastReason, if it is one of the known reasons
-            if (exception != null)
-            {
-                if (reason == RhFailFastReason.UnhandledException)
-                    errorCode = (uint)(exception.GetEETypePtr().GetHashCode());
-                else if (exception.HResult != 0)
-                    errorCode = (uint)exception.HResult;
-            }
-            else if (reason != RhFailFastReason.Unknown)
+            _crashInfo.Close();
+
+            // Try to map the failure into a HRESULT that makes sense
+            uint errorCode = exception != null ? (uint)exception.HResult : reason switch
             {
-                errorCode = (uint)reason + 0x1000; // Add something to avoid common low level exit codes
-            }
+                RhFailFastReason.EnvironmentFailFast => COR_E_APPLICATION,
+                RhFailFastReason.InternalError or
+                RhFailFastReason.ClassLibDidNotTranslateExceptionID => COR_E_EXECUTIONENGINE,
+                RhFailFastReason.UnhandledException or
+                RhFailFastReason.UnhandledExceptionFromPInvoke or
+                RhFailFastReason.UnhandledException_ExceptionDispatchNotAllowed or
+                RhFailFastReason.UnhandledException_CallerDidNotHandle => E_ACCESSDENIED,
+                _ => E_FAIL
+            };
 
-            Interop.Kernel32.RaiseFailFastException(errorCode, pExAddress, pExContext);
+#if TARGET_WINDOWS
+            Interop.Kernel32.RaiseFailFastException(errorCode, pExAddress, pExContext, _crashInfo.GetTriageBuffer(out int size), size);
 #else
             Interop.Sys.Abort();
 #endif
         }
 
-        private unsafe struct CrashDumpRecord
-        {
-            public const int CookieSize = 20;
-            public fixed byte Cookie[CookieSize];
-            public int Type;
-            public void* Data;
-            public int Length;
-        }
+        private const uint E_FAIL = 0x80004005;
```

Comment:
You can get the HResults values from src\libraries\Common\src\System\HResults.cs. (`E_ACCESSDENIED` is not there - feel free to add it.)